### DISCUSSION TITLE
carbon footprint display

### DISCUSSION THREAD

**Discussion by pierreloicq** [osmandapp/OsmAnd#11943] [Ideas]
Hello,

I think it would be useful for planet earth to display an approximated carbon footprint when we generate a route in the car mode.

**Comment by spacekookie** [comment]
There would need to be a setting for your engine's CO2 emissions in the vehicle settings but I think this is a cool idea

**Reply by Lee-Carre** [reply]
When predicting energy-use (fuel consumed, waste released, etc.), then that might do.

However, the actual result will depend heavily on how the vehicle is driven along the actual route.

Plus, other factors, like vehicle maintenance.

So, it would have to be with the caveat that it's an approximate estimate, at best.

Analysis of efficiency might be useful to inform the fuel-saving router mode.

**Comment by pierreloicq** [comment]
Do you have a page explaining the process of new features acceptation ? Is it a vote or someone will eventually decide ?
And a page describing the technologies and architecture of the app ? (I couldn't find any)
Thanks

**Reply by pierreloicq** [reply]
@spacekookie maybe you know ? Thanks

**Reply by Lee-Carre** [reply]
It's essentially down to the developers. Currently they're busy working on OpenGL rendering.

They're receptive to contributions via pull requests, though.

**Comment by Bushmills** [comment]
What additional information could such a function give, what gas station receipts or odometers couldn't already tell?

**Reply by Bushmills** [reply]
Most effective means to incorporate functions is probably by so called "Pull request".  It works like this:  the person who wished a feature to become part of the program, clones the repo, then writes the code for that feature, and then, by "Pull request", signals to the repository owner that this fork contains changes or additions he might want to look at and possibly incorporate into a branch of his own repo.

**Reply by mariush444** [reply]
I don't like the requested feature but ...
again and again ...  we have app for developers and NOT for users

maybe other idea
developer propose how much it could be cost (money and time) then users can collect money (e.g. patronate)
if the sum is collected then developer start to implemente

**Comment by Bushmills** [comment]
There's also the aspect of somehow fitting new functions into user interface. Many users complain already that the menu system is inconsistent and overly loaded - throwing in new functions isn't likely to improve the situation.

But there's a simple method to determine CO2 footprint without having to navigate through OsmAnd menus or clutter display.  For this example, I use a car with 120000 km on its odo, and an average fuel consumption of 6,5 l/100 km. Sorry if you're used to mpg - means to convert between those do exist,

produced mass of carbon dioxide (in kilograms):  120000 * 2.4 * 6.5 / 100  -> close to 19 tons.

The result take only consumed/burned fuel into account. No additional impact from fuel production, engine oil, brakes, tires etc, or transport of any of those, is considered here.
Result will be slightly higher for Diesel powered engines.

**Reply by pierreloicq** [reply]
Hello,
Indeed, everyone can compute its own carbon footprint easily but nobody do it. 
The goal of this feature is to raise awareness of a car trip footprint to all users, not just those who care already.
Let's do it in an open source app, so that other app maker will do it too ! 

It looks like the already implemented fuel-efficient mode just takes the shortest way, is it right ?

At least there is room for it : 
![Screenshot_with_CO2_small](https://user-images.githubusercontent.com/8809578/126040488-4469fa42-7447-4b57-97aa-b427c78cce85.png)

**Reply by Bushmills** [reply]
It's my belief that if a person doesn't bother to mentally hash driven kilometers to produced CO2, - using a value that person would have to enter into OsmAnd first, which is even sugarcoated by the manufacturer -, that person also won't bother about any indication of the amount of CO2 produced for a trip for which OsmAnd was used to calculate a route for. 
Those who do really care don't drive on mineral fuels, and those who are, say "aware to some extent" know how to divide driven kilometers by about vehicle specific 5 to 12. I consider this suggestion well-meant but unlikely to achieve any effect of the kind you hope for. A better place for the amount of expelled CO2 is, IMHO, printed on gas station receipts, as numbers there would  also match reality.

**Reply by pierreloicq** [reply]
I do not own any gas station so cannot change the gas station receipt.
I do not ask anyone to write this feature, just to make sure that it will be accepted if it is done.

**Reply by Bushmills** [reply]
No, you don't. But campaigning for it is still possible. This resembles a bit campaigning for adding a feature to a program.

**Reply by Lee-Carre** [reply]
> everyone can compute its own carbon footprint easily but nobody do it.

Because there are other factors involved. Applying a uni-variant analysis, where a multi-variant analysis is a more representative (of reality) model, leads nowhere productive.

Even with a rigorous analysis, that doesn't change fundamentals (at least not anytime soon).
* some number of journeys must be made
* most people don't (read: can't afford to) change their vehicle often (and a lot of energy when into manufacturing it, so getting the full service life out of it is sensible)
* some places have terrible public transport, and personal vehicles are a necessity for the basics (like buying food)

Other, broader factors are pertinent:
* (battery) EV (as opposed to fuel-cell) is hardly good for the environment, either (mining all that lithium, for starters)
* if we do nuclear sensibly, then we can have a closed-loop hydrocarbon cycle (I can elaborate, if readers would like)

People are already incentivised to use no more than they need; price & micro-economics.

I'd argue that OsmAnd (& related) helps by informing users of optimal routes. However, everything involved to yield that, uses plenty of energy and material resources.

> The goal of this feature is to raise awareness of a car trip footprint to all users, not just those who care already. Let's do it in an open source app, so that other app maker will do it too !

Hmm. The word propaganda comes to mind.

The goal of OsmAnd is to help users make their journeys, not to preach at them.

Imposing unrequested things upon users goes against so many principles of ethics related to user-respecting software. That's what developers of proprietary software do.

OSM becoming more like Google? No thanks.

I could also comment plenty on the problems of trying to inject politics into software & tech.

Libre software is, by nature, anti-authoritarian. So, I don't see how a pseudo-feature which runs counter to the goals of OsmAnd, fits with OsmAnd, and ultimately serves users.

Like has been established; for those who really care about the actual numbers, they can calculate them easily. There are apps for measuring accurate values of fuel-efficiency.

Putting my regular-user hat (with the benefit of UX knowledge & experience) on, for a moment;
* how does this suggested change help me, as a user?
* what am I supposed to do (differently than I'm already doing) with the info it might give me?
* why would I take a sub-optimal route anyway?

Ultimately, people are already pretty aware; difficult to not be, these days.

> fuel-efficient mode just takes the shortest way, is it right ?

it's not that simple.

Sometimes the fuel-efficient route, and the shortest route, will match. Not always.

Distance is but one factor.

> At least there is room for it : [Screenshot]

Many users would strongly disagree, and some would have other ideas on how display area could / should be-used.

Keep in mind that an app which is primarily about displaying a map, can't cover the map with many widgets (and everyone has their favourites, which they think should take priority). Displaying the map is of prime importance.

**Reply by Lee-Carre** [reply]
> I do not own any gas station so cannot change the gas station receipt.

Ownership isn't a requirement. Apply similar tactics;
* persuade
* encourage

Though, even if price weren't a factor (because £1.50.per litre), establishments selling fuel (or any other commodity) have (surprise) an interest in selling more of it. So, something which discourages sales (to the point of lowest need) isn't something that I imagine they'll be keen on.

**Reply by Bushmills** [reply]
> Imposing unrequested things upon users goes against so many principles of ethics related to user-respecting software. That's what developers of proprietary software do.  

The chosen method of incorporating that feature by means of a plugin allows users to enable it, if they wish, or keep it disabled. They are therefore in control of that feature showing. This isn't exactly "imposing", but more like "offering a choice", which is good, IMHO.

**Reply by Lee-Carre** [reply]
> incorporating that feature by means of a plugin allows users to enable it, if they wish, or keep it disabled. They are therefore in control of that feature showing.

Indeed. As a (default disabled) plugin, would be far less objectionable.

I don't presume to forbid it for those who do want it (which is no better). My objection is to such things being foisted upon all, regardless of their wishes / choices / preferences / ability to do anything with the info.

Ultimately, combustion emissions are a social problem, best solved by addressing that nature of it. Trying to solve a social problem, with technology, is liable to fail (and be wasted effort & resources).
Trying to use software to nag people into compliance, smacks of being Orwellian (which we already have plenty of, lately).

By all means, inform and educate the receptive, who are positioned & motivated to make whatever changes are necessary. That's not everyone, though.

**Comment by pierreloicq** [comment]
@EugeneZmeuk @Chumva  @alexey-pelykh @sonora @vshcherb
I tag you because I would love to get an official answer to my question: "would this feature be accepted if I write it ?" 
Thank you

**Reply by sonora** [reply]
Thanks for the idea!

Speaking for myself only now: My first thoughts were:
- What is the point of burning cpu cycles and battery power to inform the user that if they drove this route with an average passenger car they would produce say 820 grams of CO2, knowing that if they used an RV it would be like 3 times that, with a truck at least five times that. And with a Tesla it is more complex still unless you know how your electricity is produced.
- Does it make a difference vs. e.g announcing that this trip would burn 289 milli-liters of gas in an average passenger car, or maybe saying it would cost  you 20 cents?

I can see that there may be the point of raising awareness, but I am not convinced a navigation app would help make a sustainable difference here. It feels much like telling a jogger that during this trip it is recommended to drink x liters of water, and 3 times that if under desert conditions. It may raise an eyebrow at first, but I think this may quickly wear off or even become teaching and annoying?

The issue we have on this planet is creating awareness that CO2 production must be severely limited. But throwing around numbers like "This would cteate x kilograms of CO2" is much like saying "This would burn y kilograms of gasoline",  a total equivalent. I doubt that either statement is too helpful, because nobody has a feeling for how much of a contribution to the problem that actually is?

**Reply by vshcherb** [reply]
I would be interested to have integrated as a plugin, so once it's done we could help to wrap it up as a plugin, so it could be used.

**Reply by pierreloicq** [reply]
@sonora My idea is not to provide the comparison between vehicles. The users will choose their vehicle type and the app will provide an approximation for this type. It should not take battery and CPU, there is just a multiplication. 
We are allowed to emit ~2 tons/year to be neutral.
Expressing as a percentage of this 2 tons may be an idea, but I find it more lecturing while kg is just a fact. 

@vshcherb shipping it as a plugin will not help my goal since only people who know it exists and install it will benefit from it, right ?

**Reply by mariush444** [reply]
@pierreloicq  `only people who know it exists and install it will benefit from it`  then you decide that everyone have to go your way.
No, not everyone is child of eco-business

**Reply by pierreloicq** [reply]
Is it possible to make it as plugin that is installed by default ?

**Reply by Lee-Carre** [reply]
> We are allowed to emit ~2 tons/year to be neutral.

Tell that to Canadians & Scandinavians, especially in winter.

**Comment by pierreloicq** [comment]
Work is progressing. I have the CO2 display in the menu as I showed above. I can now start working on the menu asking for type of fuel and consumption. I plan to put it in : do prepare a route/settings/navigation settings/vehicle parameters. I think the best would be to propose some simple options by default (gazoline/diesel/electric/LPG) with averaged values and also propose a bigger choice based on the car brand and model, if I find an external service with API.

**Reply by Lee-Carre** [reply]
> an external service with API.

Part of the appeal of OsmAnd is _offline_ use. A variety of reasons, for that.

**Comment by pierreloicq** [comment]
I am almost done :-)

![work2](https://user-images.githubusercontent.com/8809578/132132638-0b331134-cede-4048-a08d-f3839ed812c6.png)
![work1](https://user-images.githubusercontent.com/8809578/132132639-8e43b011-31f5-4b93-a9d0-f92cb1ce5c65.png)
![work3](https://user-images.githubusercontent.com/8809578/132132637-22dea9ef-5add-44ca-8542-94a892bc0376.png)

**Comment by pierreloicq** [comment]
And push  https://github.com/osmandapp/OsmAnd/pull/12859 !


![Screenshot_20210926-145602](https://user-images.githubusercontent.com/8809578/134809932-50bdcc99-0e5f-4391-8250-4af8a39e8060.png)
![Screenshot_20210926-145740](https://user-images.githubusercontent.com/8809578/134809933-42bf4ba2-45dd-4619-b9de-0642c1388f4f.png)
![Screenshot_20210926-145755](https://user-images.githubusercontent.com/8809578/134809935-b9a4bb0f-2efa-4535-aebc-801b79a3033e.png)
![Screenshot_20210926-145805](https://user-images.githubusercontent.com/8809578/134809936-d8ef92c5-3062-40e1-b35e-50594255f3b0.png)
![Screenshot_20210926-145847](https://user-images.githubusercontent.com/8809578/134809937-cd6f15a0-d3a9-4c4c-93ae-a510c9fd38ad.png)
![Screenshot_20210926-145913](https://user-images.githubusercontent.com/8809578/134809938-4ce688cb-bf6e-4865-8f1e-2ea97ecad565.png)
![Screenshot_20210926-150156](https://user-images.githubusercontent.com/8809578/134809939-df31bcc0-390c-4fdb-b587-b80713b71175.png)
![Screenshot_20210926-150208](https://user-images.githubusercontent.com/8809578/134809940-1f442b99-91fd-443d-8949-e60957389ef2.png)

**Comment by okias** [comment]
Personally, I'd like to appeal to not add element to a already distracting screen (from it's goal which should be navigation).

If someone wants to have CO2 information it would be best to have it as a plugin.. (IMHO).

Preferably together with all important data as break pad and rotors wear emissions, tires emisions, emissions of vehicle construction, for EV also battery wear -> recycling impact calculations etc.

**Comment by Lee-Carre** [comment]
This might actually be an interesting basis for a routing algorithm (beyond just OsmAnd); metrics (beside distance & time) of the selected route.
If hosted along with an API, then OsmAnd could use it as an online routing provider.

However, what would be useful is to compare different modes of transport for reaching the destination. That's likely to have the largest effect on the resultant net energy-use.

**Comment by Cppugino** [comment]
It would be useful to be able to enter and save the average fuel consumption/energy consumption of a vehicle plus the approximate current fuel costs/energy costs in the navigation planning (settings). The goal would be an output of fuel costs at the planned route.

Would this be feasible in the short term?

**Reply by Bushmills** [reply]
Such a mode may even be consistent with the initial goal of the plugin, as many people seem to be more cost driven than environmental-impact driven.

**Reply by okias** [reply]
I would be personally very interested for this feature. Due to occasional carpooling, it would be nice, if the prices could be downloaded from the internet on daily basis according to nav area. So you can easily just say, how much the trip would cost.

EDIT: for the feature thou it would make sense make custom issue.

**Reply by jackyflyer** [reply]
this feature is already available on viamichelin.com

**Reply by Bushmills** [reply]
> this feature is already available on viamichelin.com

So how would one lift it out of viamichelin.com and use it with/incorporate it into OsmAnd (assuming it's GPL'ed or equivalent, or course)?

**Comment by breversa** [comment]
The French government has asked/demanded that Waze and Google Maps display the carbon footprint of the routes, among other things:
https://www.ladepeche.fr/2022/08/09/voici-pourquoi-waze-plans-et-google-maps-ne-vous-proposeront-plus-le-trajet-le-plus-court-pour-arriver-a-votre-destination-10480323.php

Not a word about non-commercial, open-source apps (to which the government seems oblivious), but I agree that it’s a nudge in the right direction. :)
